### DISCUSSION TITLE
Is it possible to format a dictionary?

### DISCUSSION THREAD

**Discussion by iliashkolyar** [axuno/SmartFormat#324] [Q&A]
Hello,
I'm using an old version (2.5.3) in which I am overriding the default braces - `Smart.Default.Parser.UseAlternativeBraces('±', '§');`

I'm using it to create dynamic JSON payloads constructed by received parameters, for example:
```
var stringToFormat = "{\"summary\": \"±body.summary§\", \"something\": \"±body.somethingElse§\"}";
var values = new {
    body = new { summary = "SUMMARY", something = "SOMETHING"} 
};

var formattedString = Smart.Format(stringToFormat, values);
```

Now this works great with nested `values` dictionary and even works properly for lists (with some workaround that I had to do because https://github.com/axuno/SmartFormat/discussions/321 didn't work in this version).

Question is, is it possible to inject a `Dictionary<string, object>` similarly to the way we add a list?
Something like this:
```
const string stringToFormat = "{\"dict\": [±field§]}";
var values = new {
    field = new Dictionary<string, object> {
        ["something"] = new { id = "123"}
    }
};
```

Currently, it just calls the `ToString` of the `Dictionary<string, object>` and I receive:
```
{"dict": [System.Collections.Generic.Dictionary`2[System.String,System.Object]]}
```

Is there some way of making the Dictionary actually format its keys and values according to their `ToString` instead of performing the `ToString` on the dictionary itself?

In the above example, I would like to receive:
```
{"dict": {"something": {"id": "123"}}}
```

Hope I made myself clear,
Thanks!

**Comment by axunonb** [comment]
The placeholder should be `field.something`.

**Reply by iliashkolyar** [reply]
In the example I posted you are indeed correct.
I rechecked my code and in my case it is:
```
const string stringToFormat = "{\"dict\": [±field.something§]}";
var values = new {
    field = new Dictionary<string, object> {
        ["something"] = new Dictionary<string, string> { ["id"] = "123" }
    }
};
```

So if the value is a `Dictionary<string, object>`, I receive the `ToString` of the dictionary.
If it's a raw object (`new { id = "123"}`), then it indeed works.

Anyway for it to work with the `Dictionary<string, object>` as well?

**Reply by axunonb** [reply]
Yes, using `field.something.id` should work. Did you try that?

**Reply by iliashkolyar** [reply]
The idea behind the code is to receive "dynamic" configuration and inject it into the formatter.
From the code point of view, I only know that `something` represents a `Dictionary<string, object>`.
I do not know what it will hold and how it is structured.
That is why I need to refer to it just as `field.something`.

**Reply by axunonb** [reply]
Without knowing what's inside it is possible to output list members.
This way you could "[synchronize](https://github.com/axuno/SmartFormat/wiki/Lists-_-ListFormatter#synchronize-2-lists)" the keys and the values of a dictionary.
It looks like you'll have to implement a custom `IFormatter` that will handle your scenario for "unknown structure".
